File: apps/android/app/src/main/java/com/signalasi/chat/AgentAttachmentPublishOrder.kt
package com.signalasi.chat

import org.json.JSONObject

/** Keeps the Signal PreKey message ahead of all later messages in a fresh session. */
internal object AgentAttachmentPublishOrder {
    data class Step(
        val attachment: AgentPreparedOutboundAttachment,
        val chunkIndex: Int?,
        val eagerChunks: Boolean = false
    ) {
        val type: String
            get() = if (chunkIndex == null) "input_attachment_manifest" else "input_attachment_chunk"

        fun payload(): JSONObject = chunkIndex?.let(attachment::chunkPayload)
            ?: attachment.manifestPayload(resume = false, eagerChunks = eagerChunks)
    }

    fun steps(attachments: List<AgentPreparedOutboundAttachment>): List<Step> = buildList {
        attachments.forEach { attachment ->
            add(Step(attachment, chunkIndex = null))
            repeat(attachment.chunkCount) { chunkIndex ->
                add(Step(attachment, chunkIndex))
            }
        }
    }

    fun initialSteps(attachments: List<AgentPreparedOutboundAttachment>): List<Step> = buildList {
        attachments.forEach { attachment ->
            val eager = attachment.sizeBytes <= EAGER_TRANSFER_BYTES
            add(Step(attachment, chunkIndex = null, eagerChunks = eager))
            if (eager) repeat(attachment.chunkCount) { chunkIndex ->
                add(Step(attachment, chunkIndex = chunkIndex))
            }
        }
    }

    private const val EAGER_TRANSFER_BYTES = 1024L * 1024L
}

File: apps/ios/GalaxySSI/AgentAttachmentPublishOrder.swift
import Foundation

enum AgentAttachmentPublishOrder {
  struct Step {
    let attachment: AgentPreparedOutboundAttachment
    let chunkIndex: Int?

    var type: String {
      chunkIndex == nil ? "input_attachment_manifest" : "input_attachment_chunk"
    }

    func payload() throws -> [String: Any] {
      if let chunkIndex {
        return try attachment.chunkPayload(index: chunkIndex)
      }
      return attachment.manifestPayload(resume: false)
    }
  }

  static func steps(_ attachments: [AgentPreparedOutboundAttachment]) -> [Step] {
    var result: [Step] = []
    for attachment in attachments {
      result.append(Step(attachment: attachment, chunkIndex: nil))
      result.append(contentsOf: (0..<attachment.chunkCount).map { index in
        Step(attachment: attachment, chunkIndex: index)
      })
    }
    return result
  }
}
